File: android/app/src/test/java/com/insituledger/app/util/RruleTest.kt
package com.insituledger.app.util

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import java.time.LocalDateTime

class RruleTest {

    companion object {
        // Every rrule string a shipped client has ever written. The TypeScript
        // twin (frontend/src/lib/rrule.test.ts) has this same table verbatim —
        // keep them in step.
        private val LEGACY_WIRE_STRINGS = listOf(
            "FREQ=DAILY",
            "FREQ=WEEKLY",
            "FREQ=WEEKLY;INTERVAL=2",
            "FREQ=MONTHLY",
            "FREQ=MONTHLY;INTERVAL=3",
            "FREQ=YEARLY"
        )
    }

    // --- parse ---

    @Test
    fun parsesEveryPreset() {
        assertEquals(ParsedRrule(Freq.DAILY, 1, ""), Rrule.parse("FREQ=DAILY"))
        assertEquals(ParsedRrule(Freq.WEEKLY, 1, ""), Rrule.parse("FREQ=WEEKLY"))
        assertEquals(ParsedRrule(Freq.WEEKLY, 2, ""), Rrule.parse("FREQ=WEEKLY;INTERVAL=2"))
        assertEquals(ParsedRrule(Freq.MONTHLY, 1, ""), Rrule.parse("FREQ=MONTHLY"))
        assertEquals(ParsedRrule(Freq.MONTHLY, 3, ""), Rrule.parse("FREQ=MONTHLY;INTERVAL=3"))
        assertEquals(ParsedRrule(Freq.YEARLY, 1, ""), Rrule.parse("FREQ=YEARLY"))
    }

    @Test
    fun parsesCustomInterval() {
        assertEquals(ParsedRrule(Freq.MONTHLY, 2, ""), Rrule.parse("FREQ=MONTHLY;INTERVAL=2"))
        assertEquals(45, Rrule.parse("FREQ=DAILY;INTERVAL=45").interval)
    }

    @Test
    fun normalizesEveryAcceptedUntilShape() {
        assertEquals("2026-12-31", Rrule.parse("FREQ=MONTHLY;UNTIL=20261231T235959Z").until)
        assertEquals("2026-12-31", Rrule.parse("FREQ=MONTHLY;UNTIL=20261231T235959").until)
        assertEquals("2026-12-31", Rrule.parse("FREQ=MONTHLY;UNTIL=20261231").until)
        assertEquals("2026-12-31", Rrule.parse("FREQ=MONTHLY;UNTIL=2026-12-31").until)
        assertEquals("2026-12-31", Rrule.parse("FREQ=MONTHLY;UNTIL=2026-12-31T23:59:59").until)
    }

    @Test
    fun ignoresUnrecognizedUntil() {
        assertEquals("", Rrule.parse("FREQ=MONTHLY;UNTIL=nonsense").until)
    }

    @Test
    fun parseIsOrderIndependent() {
        assertEquals(ParsedRrule(Freq.WEEKLY, 2, ""), Rrule.parse("INTERVAL=2;FREQ=WEEKLY"))
        assertEquals(
            ParsedRrule(Freq.DAILY, 4, "2026-12-31"),
            Rrule.parse("UNTIL=20261231T235959Z;FREQ=DAILY;INTERVAL=4")
        )
    }

    @Test
    fun fallsBackToIntervalOneForBadValues() {
        assertEquals(1, Rrule.parse("FREQ=DAILY;INTERVAL=0").interval)
        assertEquals(1, Rrule.parse("FREQ=DAILY;INTERVAL=abc").interval)
        assertEquals(1, Rrule.parse("FREQ=DAILY;INTERVAL=-2").interval)
        assertEquals(1, Rrule.parse("FREQ=DAILY;INTERVAL=").interval)
    }

    @Test
    fun skipsUnknownKeysAndMalformedSegments() {
        assertEquals(
            ParsedRrule(Freq.WEEKLY, 3, ""),
            Rrule.parse("FREQ=WEEKLY;BYDAY=MO;;INTERVAL=3;junk")
        )
    }

    @Test
    fun fallsBackToMonthlyForMissingOrGarbageFreq() {
        assertEquals(ParsedRrule(Freq.MONTHLY, 1, ""), Rrule.parse(""))
        assertEquals(ParsedRrule(Freq.MONTHLY, 2, ""), Rrule.parse("INTERVAL=2"))
        assertEquals(ParsedRrule(Freq.MONTHLY, 1, ""), Rrule.parse("FREQ=HOURLY"))
        assertEquals(ParsedRrule(Freq.MONTHLY, 1, ""), Rrule.parse("total garbage"))
    }

    // --- build ---

    @Test
    fun buildMatchesLegacyWireStrings() {
        // A refactor must not silently change the wire format: these are the
        // exact strings every shipped client has written since v1.0.
        assertEquals(LEGACY_WIRE_STRINGS, Rrule.PRESETS.map { Rrule.build(it.freq, it.interval) })
    }

    @Test
    fun buildOmitsIntervalOne() {
        // Non-negotiable: pre-v1.33.0 clients match the rrule by exact string,
        // so "FREQ=WEEKLY;INTERVAL=1" would degrade to "monthly" and overwrite
        // the real recurrence on save.
        assertEquals("FREQ=WEEKLY", Rrule.build(Freq.WEEKLY, 1))
        assertEquals("FREQ=DAILY", Rrule.build(Freq.DAILY, 1))
    }

    @Test
    fun buildEmitsCustomInterval() {
        assertEquals("FREQ=MONTHLY;INTERVAL=2", Rrule.build(Freq.MONTHLY, 2))
        assertEquals("FREQ=DAILY;INTERVAL=45", Rrule.build(Freq.DAILY, 45))
        assertEquals("FREQ=YEARLY;INTERVAL=999", Rrule.build(Freq.YEARLY, Rrule.MAX_INTERVAL))
    }

    @Test
    fun buildKeepsKeyOrderAndCompactUntil() {
        assertEquals("FREQ=MONTHLY;UNTIL=20261231T235959Z", Rrule.build(Freq.MONTHLY, 1, "2026-12-31"))
        assertEquals(
            "FREQ=WEEKLY;INTERVAL=3;UNTIL=20261231T235959Z",
            Rrule.build(Freq.WEEKLY, 3, "2026-12-31")
        )
        assertEquals("FREQ=MONTHLY", Rrule.build(Freq.MONTHLY, 1, ""))
    }

    @Test
    fun roundTripsEveryKnownString() {
        // The regression guard for the data-loss bug: before v1.33.0 the form
        // reverse-mapped an rrule by exact equality, so anything unrecognized
        // became "monthly" and saving silently rewrote the real recurrence.
        val cases = LEGACY_WIRE_STRINGS +
            LEGACY_WIRE_STRINGS.map { "$it;UNTIL=20261231T235959Z" } +
            listOf(
                "FREQ=MONTHLY;INTERVAL=2",
                "FREQ=DAILY;INTERVAL=45",
                "FREQ=WEEKLY;INTERVAL=3;UNTIL=20261231T235959Z",
                "FREQ=YEARLY;INTERVAL=999"
            )
        for (s in cases) {
            val p = Rrule.parse(s)
            assertEquals(s, Rrule.build(p.freq, p.interval, p.until))
        }
    }

    // --- label / presetKey ---

    @Test
    fun labelsEveryPreset() {
        assertEquals("Daily", Rrule.label("FREQ=DAILY"))
        assertEquals("Weekly", Rrule.label("FREQ=WEEKLY"))
        assertEquals("Biweekly", Rrule.label("FREQ=WEEKLY;INTERVAL=2"))
        assertEquals("Monthly", Rrule.label("FREQ=MONTHLY"))
        assertEquals("Quarterly", Rrule.label("FREQ=MONTHLY;INTERVAL=3"))
        assertEquals("Yearly", Rrule.label("FREQ=YEARLY"))
    }

    @Test
    fun labelIgnoresUntilWhenPresetMatching() {
        // Pre-v1.33.0 this leaked the raw string for every ending schedule.
        assertEquals("Biweekly", Rrule.label("FREQ=WEEKLY;INTERVAL=2;UNTIL=20261231T235959Z"))
        assertEquals("Daily", Rrule.label("FREQ=DAILY;UNTIL=20261231T235959Z"))
    }

    @Test
    fun labelsCustomIntervals() {
        assertEquals("Every 2 months", Rrule.label("FREQ=MONTHLY;INTERVAL=2"))
        assertEquals("Every 10 days", Rrule.label("FREQ=DAILY;INTERVAL=10"))
        assertEquals("Every 5 weeks", Rrule.label("FREQ=WEEKLY;INTERVAL=5"))
        assertEquals("Every 2 years", Rrule.label("FREQ=YEARLY;INTERVAL=2"))
    }

    @Test
    fun labelNeverLeaksRawString() {
        assertEquals("Daily", Rrule.label("FREQ=DAILY;INTERVAL=1"))
        assertEquals("Monthly", Rrule.label("total garbage"))
    }

    @Test
    fun presetKeyResolvesPairs() {
        assertEquals("daily", Rrule.presetKey(Freq.DAILY, 1))
        assertEquals("weekly", Rrule.presetKey(Freq.WEEKLY, 1))
        assertEquals("biweekly", Rrule.presetKey(Freq.WEEKLY, 2))
        assertEquals("monthly", Rrule.presetKey(Freq.MONTHLY, 1))
        assertEquals("quarterly", Rrule.presetKey(Freq.MONTHLY, 3))
        assertEquals("yearly", Rrule.presetKey(Freq.YEARLY, 1))
    }

    @Test
    fun presetKeyReturnsCustomOtherwise() {
        assertEquals(Rrule.CUSTOM_KEY, Rrule.presetKey(Freq.MONTHLY, 2))
        assertEquals(Rrule.CUSTOM_KEY, Rrule.presetKey(Freq.WEEKLY, 3))
        assertEquals(Rrule.CUSTOM_KEY, Rrule.presetKey(Freq.YEARLY, 2))
        assertEquals(Rrule.CUSTOM_KEY, Rrule.presetKey(Freq.DAILY, 45))
    }

    @Test
    fun unitLabelIsSingularForOne() {
        assertEquals("day", Rrule.unitLabel(Freq.DAILY, 1))
        assertEquals("days", Rrule.unitLabel(Freq.DAILY, 2))
        assertEquals("week", Rrule.unitLabel(Freq.WEEKLY, 1))
        assertEquals("weeks", Rrule.unitLabel(Freq.WEEKLY, 3))
        assertEquals("month", Rrule.unitLabel(Freq.MONTHLY, 1))
        assertEquals("months", Rrule.unitLabel(Freq.MONTHLY, 2))
        assertEquals("year", Rrule.unitLabel(Freq.YEARLY, 1))
        assertEquals("years", Rrule.unitLabel(Freq.YEARLY, 4))
    }

    // --- advance ---

    @Test
    fun advanceClampsMonthEnd() {
        // Same table as TestAdvanceDateMonthEndClamps in the Go scheduler —
        // this is the cross-platform parity check.
        val cases = listOf(
            Triple("2026-01-31T09:00", Freq.MONTHLY to 1, "2026-02-28T09:00"),
            Triple("2024-01-31T09:00", Freq.MONTHLY to 1, "2024-02-29T09:00"),
            Triple("2026-01-31T09:00", Freq.MONTHLY to 3, "2026-04-30T09:00"),
            Triple("2026-03-31T09:00", Freq.MONTHLY to 1, "2026-04-30T09:00"),
            Triple("2026-05-15T09:00", Freq.MONTHLY to 1, "2026-06-15T09:00"),
            Triple("2026-12-31T09:00", Freq.MONTHLY to 1, "2027-01-31T09:00"),
            Triple("2026-11-30T09:00", Freq.MONTHLY to 2, "2027-01-30T09:00")
        )
        for ((input, rule, want) in cases) {
            val got = Rrule.advance(LocalDateTime.parse(input), rule.first, rule.second)
            assertEquals(input, LocalDateTime.parse(want), got)
        }
    }

    @Test
    fun advanceClampsYearlyLeapDay() {
        assertEquals(
            LocalDateTime.parse("2025-02-28T09:00"),
            Rrule.advance(LocalDateTime.parse("2024-02-29T09:00"), Freq.YEARLY, 1)
        )
        assertEquals(
            LocalDateTime.parse("2028-02-29T09:00"),
            Rrule.advance(LocalDateTime.parse("2024-02-29T09:00"), Freq.YEARLY, 4)
        )
    }

    @Test
    fun advanceClampDrifts() {
        // Accepted behaviour, locked so nobody "fixes" it by accident: the clamp
        // is sticky, so Jan 31 → Feb 28 → Mar 28, not back to Mar 31. True RFC
        // 5545 semantics would need an anchor day-of-month column. See TODO.
        val feb = Rrule.advance(LocalDateTime.parse("2026-01-31T09:00"), Freq.MONTHLY, 1)
        assertEquals(LocalDateTime.parse("2026-02-28T09:00"), feb)
        assertEquals(LocalDateTime.parse("2026-03-28T09:00"), Rrule.advance(feb, Freq.MONTHLY, 1))
    }

    @Test
    fun advanceDateHonoursInterval() {
        // Regression for the form preview, which used to advance by the UI key
        // and so ignored INTERVAL entirely.
        assertEquals("2026-01-11", Rrule.advanceDate("2026-01-01", "FREQ=DAILY;INTERVAL=10").first)
        assertEquals("2026-02-05", Rrule.advanceDate("2026-01-01", "FREQ=WEEKLY;INTERVAL=5").first)
        assertEquals("2026-08-01", Rrule.advanceDate("2026-01-01", "FREQ=MONTHLY;INTERVAL=7").first)
        assertEquals("2028-01-01", Rrule.advanceDate("2026-01-01", "FREQ=YEARLY;INTERVAL=2").first)
    }

    @Test
    fun advanceDatePreservesDateOnlyShape() {
        val (next, _) = Rrule.advanceDate("2026-01-31", "FREQ=MONTHLY")
        assertEquals("2026-02-28", next)
    }

    @Test
    fun advanceDateEmitsOffsetForDatetimeInput() {
        val (next, _) = Rrule.advanceDate("2026-01-31T09:00", "FREQ=MONTHLY")
        // RFC3339 with the system zone's offset; assert the local part and that
        // an offset (or Z) is present, so the test is TZ-independent.
        assertTrue(next, next.startsWith("2026-02-28T09:00"))
        assertTrue(next, next.indexOf('+', 11) >= 0 || next.indexOf('-', 11) >= 0 || next.endsWith("Z"))
    }

    @Test
    fun advanceDateUntilBoundary() {
        // Landing exactly on UNTIL keeps the schedule alive; passing it ends it.
        val (onUntil, past1) = Rrule.advanceDate("2026-11-30", "FREQ=MONTHLY;UNTIL=20261231T235959Z")
        assertEquals("2026-12-30", onUntil)
        assertFalse(past1)

        val (_, past2) = Rrule.advanceDate("2026-12-30", "FREQ=MONTHLY;UNTIL=20261231T235959Z")
        assertTrue(past2)
    }

    @Test
    fun advanceDateWithoutUntilNeverExpires() {
        val (_, pastUntil) = Rrule.advanceDate("2026-01-01", "FREQ=DAILY")
        assertFalse(pastUntil)
    }

    @Test
    fun parseUntilRejectsGarbage() {
        assertNull(Rrule.parseUntil("nonsense"))
        assertEquals(LocalDateTime.parse("2026-12-31T23:59:59"), Rrule.parseUntil("20261231"))
    }
}

File: frontend/src/lib/rrule.test.ts
import { describe, it, expect } from 'vitest';
import {
	MAX_INTERVAL,
	PRESETS,
	buildRrule,
	parseRrule,
	presetKeyFor,
	rruleLabel,
	unitLabel
} from './rrule';

// Every rrule string a shipped client has ever written. The Kotlin twin
// (android/…/util/Rrule.kt) has this same table verbatim — keep them in step.
const LEGACY_WIRE_STRINGS = [
	'FREQ=DAILY',
	'FREQ=WEEKLY',
	'FREQ=WEEKLY;INTERVAL=2',
	'FREQ=MONTHLY',
	'FREQ=MONTHLY;INTERVAL=3',
	'FREQ=YEARLY'
];

describe('parseRrule', () => {
	it('parses every preset', () => {
		expect(parseRrule('FREQ=DAILY')).toEqual({ freq: 'DAILY', interval: 1, until: '' });
		expect(parseRrule('FREQ=WEEKLY')).toEqual({ freq: 'WEEKLY', interval: 1, until: '' });
		expect(parseRrule('FREQ=WEEKLY;INTERVAL=2')).toEqual({
			freq: 'WEEKLY',
			interval: 2,
			until: ''
		});
		expect(parseRrule('FREQ=MONTHLY')).toEqual({ freq: 'MONTHLY', interval: 1, until: '' });
		expect(parseRrule('FREQ=MONTHLY;INTERVAL=3')).toEqual({
			freq: 'MONTHLY',
			interval: 3,
			until: ''
		});
		expect(parseRrule('FREQ=YEARLY')).toEqual({ freq: 'YEARLY', interval: 1, until: '' });
	});
	it('parses a custom interval', () => {
		expect(parseRrule('FREQ=MONTHLY;INTERVAL=2')).toEqual({
			freq: 'MONTHLY',
			interval: 2,
			until: ''
		});
		expect(parseRrule('FREQ=DAILY;INTERVAL=45').interval).toBe(45);
	});
	it('normalizes every accepted UNTIL shape', () => {
		expect(parseRrule('FREQ=MONTHLY;UNTIL=20261231T235959Z').until).toBe('2026-12-31');
		expect(parseRrule('FREQ=MONTHLY;UNTIL=20261231T235959').until).toBe('2026-12-31');
		expect(parseRrule('FREQ=MONTHLY;UNTIL=20261231').until).toBe('2026-12-31');
		expect(parseRrule('FREQ=MONTHLY;UNTIL=2026-12-31').until).toBe('2026-12-31');
		expect(parseRrule('FREQ=MONTHLY;UNTIL=2026-12-31T23:59:59').until).toBe('2026-12-31');
	});
	it('ignores an unrecognized UNTIL rather than emitting garbage', () => {
		expect(parseRrule('FREQ=MONTHLY;UNTIL=nonsense').until).toBe('');
	});
	it('is order-independent', () => {
		expect(parseRrule('INTERVAL=2;FREQ=WEEKLY')).toEqual({
			freq: 'WEEKLY',
			interval: 2,
			until: ''
		});
		expect(parseRrule('UNTIL=20261231T235959Z;FREQ=DAILY;INTERVAL=4')).toEqual({
			freq: 'DAILY',
			interval: 4,
			until: '2026-12-31'
		});
	});
	it('falls back to interval 1 for non-positive or unparseable values', () => {
		expect(parseRrule('FREQ=DAILY;INTERVAL=0').interval).toBe(1);
		expect(parseRrule('FREQ=DAILY;INTERVAL=abc').interval).toBe(1);
		expect(parseRrule('FREQ=DAILY;INTERVAL=-2').interval).toBe(1);
		expect(parseRrule('FREQ=DAILY;INTERVAL=').interval).toBe(1);
	});
	it('skips unknown keys and malformed segments', () => {
		expect(parseRrule('FREQ=WEEKLY;BYDAY=MO;;INTERVAL=3;junk')).toEqual({
			freq: 'WEEKLY',
			interval: 3,
			until: ''
		});
	});
	it('falls back to monthly for missing, unknown or garbage input', () => {
		expect(parseRrule('')).toEqual({ freq: 'MONTHLY', interval: 1, until: '' });
		expect(parseRrule('INTERVAL=2')).toEqual({ freq: 'MONTHLY', interval: 2, until: '' });
		expect(parseRrule('FREQ=HOURLY')).toEqual({ freq: 'MONTHLY', interval: 1, until: '' });
		expect(parseRrule('total garbage')).toEqual({ freq: 'MONTHLY', interval: 1, until: '' });
	});
});

describe('buildRrule', () => {
	it('emits the legacy wire string for every preset', () => {
		const built = PRESETS.map(p => buildRrule(p.freq, p.interval));
		expect(built).toEqual(LEGACY_WIRE_STRINGS);
	});
	it('omits INTERVAL when it is 1', () => {
		// Non-negotiable: pre-v1.22.0 clients match the rrule by exact string,
		// so `FREQ=WEEKLY;INTERVAL=1` would degrade to "monthly" and overwrite
		// the real recurrence on save.
		expect(buildRrule('WEEKLY', 1)).toBe('FREQ=WEEKLY');
		expect(buildRrule('DAILY', 1)).toBe('FREQ=DAILY');
	});
	it('emits INTERVAL for custom values', () => {
		expect(buildRrule('MONTHLY', 2)).toBe('FREQ=MONTHLY;INTERVAL=2');
		expect(buildRrule('DAILY', 45)).toBe('FREQ=DAILY;INTERVAL=45');
		expect(buildRrule('YEARLY', MAX_INTERVAL)).toBe(`FREQ=YEARLY;INTERVAL=${MAX_INTERVAL}`);
	});
	it('appends UNTIL in the compact end-of-day form', () => {
		expect(buildRrule('MONTHLY', 1, '2026-12-31')).toBe('FREQ=MONTHLY;UNTIL=20261231T235959Z');
	});
	it('keeps key order FREQ;INTERVAL;UNTIL', () => {
		expect(buildRrule('WEEKLY', 3, '2026-12-31')).toBe(
			'FREQ=WEEKLY;INTERVAL=3;UNTIL=20261231T235959Z'
		);
	});
	it('ignores an empty until', () => {
		expect(buildRrule('MONTHLY', 1, '')).toBe('FREQ=MONTHLY');
	});
});

describe('round trip', () => {
	// The regression guard for the data-loss bug: before v1.22.0 the form
	// reverse-mapped an rrule by exact equality, so anything unrecognized became
	// "monthly" and saving silently rewrote the user's real recurrence.
	const cases = [
		...LEGACY_WIRE_STRINGS,
		...LEGACY_WIRE_STRINGS.map(s => `${s};UNTIL=20261231T235959Z`),
		'FREQ=MONTHLY;INTERVAL=2',
		'FREQ=DAILY;INTERVAL=45',
		'FREQ=WEEKLY;INTERVAL=3;UNTIL=20261231T235959Z',
		`FREQ=YEARLY;INTERVAL=${MAX_INTERVAL}`
	];
	for (const s of cases) {
		it(`preserves ${s}`, () => {
			const p = parseRrule(s);
			expect(buildRrule(p.freq, p.interval, p.until)).toBe(s);
		});
	}
});

describe('rruleLabel', () => {
	it('labels every preset', () => {
		expect(rruleLabel('FREQ=DAILY')).toBe('Daily');
		expect(rruleLabel('FREQ=WEEKLY')).toBe('Weekly');
		expect(rruleLabel('FREQ=WEEKLY;INTERVAL=2')).toBe('Biweekly');
		expect(rruleLabel('FREQ=MONTHLY')).toBe('Monthly');
		expect(rruleLabel('FREQ=MONTHLY;INTERVAL=3')).toBe('Quarterly');
		expect(rruleLabel('FREQ=YEARLY')).toBe('Yearly');
	});
	it('ignores UNTIL when preset-matching', () => {
		// Pre-v1.22.0 this leaked the raw string for every ending schedule.
		expect(rruleLabel('FREQ=WEEKLY;INTERVAL=2;UNTIL=20261231T235959Z')).toBe('Biweekly');
		expect(rruleLabel('FREQ=DAILY;UNTIL=20261231T235959Z')).toBe('Daily');
	});
	it('describes custom intervals', () => {
		expect(rruleLabel('FREQ=MONTHLY;INTERVAL=2')).toBe('Every 2 months');
		expect(rruleLabel('FREQ=DAILY;INTERVAL=10')).toBe('Every 10 days');
		expect(rruleLabel('FREQ=WEEKLY;INTERVAL=5')).toBe('Every 5 weeks');
		expect(rruleLabel('FREQ=YEARLY;INTERVAL=2')).toBe('Every 2 years');
	});
	it('treats a redundant INTERVAL=1 as the preset', () => {
		expect(rruleLabel('FREQ=DAILY;INTERVAL=1')).toBe('Daily');
	});
	it('never leaks a raw string for garbage input', () => {
		expect(rruleLabel('total garbage')).toBe('Monthly');
	});
});

describe('presetKeyFor', () => {
	it('resolves preset pairs', () => {
		expect(presetKeyFor('DAILY', 1)).toBe('daily');
		expect(presetKeyFor('WEEKLY', 1)).toBe('weekly');
		expect(presetKeyFor('WEEKLY', 2)).toBe('biweekly');
		expect(presetKeyFor('MONTHLY', 1)).toBe('monthly');
		expect(presetKeyFor('MONTHLY', 3)).toBe('quarterly');
		expect(presetKeyFor('YEARLY', 1)).toBe('yearly');
	});
	it('returns custom for everything else', () => {
		expect(presetKeyFor('MONTHLY', 2)).toBe('custom');
		expect(presetKeyFor('WEEKLY', 3)).toBe('custom');
		expect(presetKeyFor('YEARLY', 2)).toBe('custom');
		expect(presetKeyFor('DAILY', 45)).toBe('custom');
	});
});

describe('unitLabel', () => {
	it('is singular for 1 and plural otherwise', () => {
		expect(unitLabel('DAILY', 1)).toBe('day');
		expect(unitLabel('DAILY', 2)).toBe('days');
		expect(unitLabel('WEEKLY', 1)).toBe('week');
		expect(unitLabel('WEEKLY', 3)).toBe('weeks');
		expect(unitLabel('MONTHLY', 1)).toBe('month');
		expect(unitLabel('MONTHLY', 2)).toBe('months');
		expect(unitLabel('YEARLY', 1)).toBe('year');
		expect(unitLabel('YEARLY', 4)).toBe('years');
	});
});
